### PR TITLE
Add TemplateAstVisitor, parse *ngFor micro expressions

### PR DESCRIPTION
**Pull request by matanlurey**
Branched from https://github.com/dart-lang/angular_ast/pull/11, which I'm waiting on review for.

### CODE REVIEW

File: lib/src/ast/attribute.dart [hunk 1]
```diff
@@ -32,8 +33,17 @@ abstract class AttributeAst implements TemplateAst {
   ]) = _ParsedAttributeAst;
 
   @override
-  bool operator ==(Object o) =>
-      o is AttributeAst && name == o.name && value == o.value;
+  /*=R*/ accept/*<R, C>*/(TemplateAstVisitor/*<R, C>*/ visitor, [C context]) {
```
Comment (MichaelRFairhurst) [inline lib/src/ast/attribute.dart:36]: is C supposed to still be here?

File: lib/src/ast/comment.dart [hunk 2]
```diff
@@ -27,6 +28,11 @@ abstract class CommentAst implements StandaloneTemplateAst {
   ) = _ParsedCommentAst;
 
   @override
+  /*=R*/ accept/*<R, C>*/(TemplateAstVisitor/*<R, C>*/ visitor, [C context]) {
```
Comment (MichaelRFairhurst) [inline lib/src/ast/comment.dart:31]: Same comment on `C`

File: lib/src/ast/expression.dart [hunk 3]
```diff
@@ -2,21 +2,41 @@
...
+
   @override
-  bool operator ==(Object o) =>
-      o is ExpressionAst && o.expression == expression;
+  bool operator ==(Object o) {
+    if (o is ExpressionAst) {
+      return o.expression.toSource() == expression.toSource();
+    }
+    return false;
+  }
```
Comment (MichaelRFairhurst) [inline lib/src/ast/expression.dart:39]: Could be useful to add to the visitor as well (for instance, we'll probably put a scope of local variables on each of these, and then run a visitor that just resolves every Expression. Might not want to have to implement visitMustache, visitInput, etc etc just to get at all the Exrpessions).

File: lib/src/ast/sugar/banana.dart [hunk 4]
```diff
@@ -35,6 +36,11 @@ abstract class BananaAst implements TemplateAst {
   ) = _ParsedBananaAst;
 
   @override
+  /*=R*/ accept/*<R, C>*/(TemplateAstVisitor/*<R, C>*/ visitor, [C context]) {
+    throw new UnimplementedError();
```
Comment (MichaelRFairhurst) [inline lib/src/ast/sugar/banana.dart:40]: Works for now but I'm hoping this changes! :)

File: lib/src/ast/sugar/star.dart [hunk 5]
```diff
@@ -36,6 +37,11 @@ abstract class StarAst implements TemplateAst {
   ]) = _ParsedStarAst;
 
   @override
+  /*=R*/ accept/*<R, C>*/(TemplateAstVisitor/*<R, C>*/ visitor, [C context]) {
+    throw new UnimplementedError();
```
Comment (MichaelRFairhurst) [inline lib/src/ast/sugar/star.dart:41]: same comment as Banana

File: lib/src/expression/micro.dart [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  String directive, {
+  bool deSugarPipes: true,
+  String sourceUrl: '/test.dart',
+}) {
+  final assignments = <ReferenceAst>[];
+  final properties = <PropertyAst>[];
+  final parts = expression.split(';');
+
+  // Parse the first statement, which is special.
+  if (!parts[0].startsWith('let')) {
```
Comment (MichaelRFairhurst) [inline lib/src/expression/micro.dart:37]: What about ngIf? doesn't have to start with let

File: lib/src/expression/micro.dart [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  }
+  var parse = parts[0].split(' ');
+  assignments.add(new ReferenceAst(parse[1]));
+  var name = parse[2];
+  name = directive + name[0].toUpperCase() + name.substring(1);
+  properties.add(
+    new PropertyAst(
+      name,
+      new ExpressionAst.parse(
+        parse[3],
```
Comment (MichaelRFairhurst) [inline lib/src/expression/micro.dart:48]: could contain spaces, ie `let x of myMethod(arg1, arg2)`

Also, see work here that @mk13 has done related to trailing expressions, which you'll need to check: https://github.com/dart-lang/angular_analyzer_plugin/pull/94/files#diff-dd70e04facf5d905f5ff348ad54053f3R987

File: lib/src/expression/micro.dart [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+  // Parse remaining statements, if any.
+  for (var part in parts.skip(1)) {
+    part = part.trim();
+    if (part.startsWith('let')) {
+      var name = part.substring(4, part.indexOf('='));
+      var value = part.substring(part.indexOf('=') + 1);
+      assignments.add(new ReferenceAst(name.trim(), value.trim()));
+    } else {
+      var subParts = part.split(':');
```
Comment (MichaelRFairhurst) [inline lib/src/expression/micro.dart:63]: I think `:` is allowed more than just here, ie `let x of: items`.

Also, I think `let x of items` is equivalent to `let x; of items`

In general I think the grammar is pretty simple:

```
Star := '*' name '=' '"' (expr | MicroExprList) '"'
MicroExprList = (LetExpr | BindExpr)+
LetExpr := 'let' id ('=' id)? ';'?
BindExpr := (id ':'?)? expr ';'?
```

Even simpler when you think of `*x="y"` as sugar for `template="x y"`

```
Template := 'template' '=' '"' MicroExprList '"'
```

File: lib/src/visitor.dart [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+///
+/// An implementation may return element [R], and optionally use [C] as context.
+abstract class TemplateAstVisitor<R, C> {
+  const TemplateAstVisitor();
+
+  /// Visits an [astNode] tree, passing along [context] if given.
+  ///
+  /// Based on the subtype other `visit` methods are called.
+  R visit(TemplateAst astNode, [C context]) {
+    if (astNode is AttributeAst) {
```
Comment (MichaelRFairhurst) [inline lib/src/visitor.dart:19]: This should just call astNode.accept(this, context), no?

**Review by MichaelRFairhurst — APPROVED** [review]
Aside from trailing expressions and some lack of a spec to conform to in terms of how exactly star syntax works, this looks good.

File: lib/src/ast/attribute.dart (hunk 1 above)
Comment (matanlurey) [reply lib/src/ast/attribute.dart:36]: Nope, will fix in #14.

File: lib/src/ast/expression.dart (hunk 3 above)
Comment (matanlurey) [reply lib/src/ast/expression.dart:39]: Ah good idea, added to #14.

File: lib/src/expression/micro.dart (hunk 6 above)
Comment (matanlurey) [reply lib/src/expression/micro.dart:37]: AFAIK ngIf is always bound to straight expression, right?

File: lib/src/expression/micro.dart (hunk 7 above)
Comment (matanlurey) [reply lib/src/expression/micro.dart:48]: Thanks! Added to #14

File: lib/src/expression/micro.dart (hunk 6 above)
Comment (MichaelRFairhurst) [reply lib/src/expression/micro.dart:37]: Ah, I see, yeah, you check that later.

I think still technically `*ngFor="of items; let item"` is valid

Comment (MichaelRFairhurst) [reply lib/src/expression/micro.dart:37]: But could be wrong

File: lib/src/expression/micro.dart (hunk 8 above)
Comment (matanlurey) [reply lib/src/expression/micro.dart:63]: Oh man, I had no idea that's what this even meant...

Then I need a much better micro expression parser. See #15

File: lib/src/visitor.dart (hunk 9 above)
Comment (matanlurey) [reply lib/src/visitor.dart:19]: Yep, that didn't exist before. Filed in #14.

**Pull request merged**